### ISSUE TITLE
Failing assertions triggered during sync in Fenix

### ISSUE THREAD

**Issue by rfk** [mozilla/application-services#4153]
We've recently started getting some crash reports from Fenix along the lines of:

```
[INFO] ffi_support::error - Caught a panic calling rust code: "assertion failed: `(left != right)`\n  left: `Guid(\"6a7a381c91ca\")`,\n right: `Guid(\"6a7a381c91ca\")`"
```

See e.g. [this Sentry report](https://sentry.prod.mozaws.net/operations/firefox-nightly/issues/8822116/), which originates in a call to `sync_manager_sync`.

The error message here reads like a failing `assert_ne!` call. I had a look through our code for instances of this outside of tests, and I found three in the autofill component:

* One here in [`apply_incoming_action`](https://github.com/mozilla/application-services/blob/e67c3aab466ac833a14a5113e88bc4e720976371/components/autofill/src/sync/mod.rs#L380)
* One here in [`plan_incoming`](https://github.com/mozilla/application-services/blob/e67c3aab466ac833a14a5113e88bc4e720976371/components/autofill/src/sync/mod.rs#L336)
* One here in [`common_change_guid`](https://github.com/mozilla/application-services/blob/e67c3aab466ac833a14a5113e88bc4e720976371/components/autofill/src/sync/common.rs#L82)

There may be others in other components, but given the recency of the spike in errors, autofill seems like a good place to start. We need to see if we can reproduce conditions that trigger these assertions, or convince ourselves that they're fine and the problem is coming from elsewhere.

Unfortunately, panics during a sync seem to have a bad habit of poisoning the store Mutex for the `tabs` and `logins` stores, potentially leading to other user-visible crashes like #4035.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/SYNC-2231)
┆Sprint: SYNC - 2021-06-11

**Comment by rfk** [comment]
@mhammond found what we believe to be the cause of this, fix in https://github.com/mozilla/application-services/pull/4154

**Comment by mhammond** [comment]
Getting confirmation this was fixed by #4154

**Comment by rfk** [comment]
I don't know why this was re-opened - automation error maybe? I'm going to close it, but please re-open if there is more work to do here.
